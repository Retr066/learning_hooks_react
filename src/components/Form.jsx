import React, { useState, useRef } from "react";
import useForm from "../hooks/useForm";
import Alert from "./Alert";

export default function Form({ dispatch }) {
  const [formValues, handleInputsChange, resetForm] = useForm({
    id: new Date().getTime(),
    title: "",
    desc: "",
    completed: false,
  });
  const [stateAlert, setViewAlert] = useState({
    viewAlert: false,
    messageAlert: "Fill inputs",
    color: "dark",
  });
  const inputRef = useRef(null);
  const { viewAlert, messageAlert, color } = stateAlert;
  const { title, desc } = formValues;

  const handleFormSubmit = (e) => {
    e.preventDefault();
    if (title.trim().length <= 1 || desc.trim().length <= 1) {
      setViewAlert({
        ...stateAlert,
        viewAlert: true,
      });
      return;
    }

    const action = {
      type: "add",
      payload: formValues,
    };
    dispatch(action);
    setViewAlert({
      ...stateAlert,
      viewAlert: false,
    });
    resetForm();
  };

  const something = (event) => {
    if (event.keyCode === 13) {
      handleFormSubmit(event);
      inputRef.current.focus();
    }
  };

  return (
    <>
      <form onSubmit={handleFormSubmit}>
        {viewAlert && <Alert text={messageAlert} color={color} />}
        <div className="mb-3">
          <label htmlFor="exampleFormControlInput1" className="form-label">
            Title
          </label>
          <input
            ref={inputRef}
            type="text"
            className="form-control"
            value={title}
            name="title"
            onChange={handleInputsChange}
            placeholder="Ingrese su nombre completo"
            autoComplete="off"
          />
        </div>

        <div className="mb-3">
          <label htmlFor="exampleFormControlTextarea1" className="form-label">
            Description
          </label>
          <textarea
            className="form-control"
            value={desc}
            name="desc"
            rows="3"
            onChange={handleInputsChange}
            onKeyDown={(e) => something(e)}
            autoComplete="off"
          ></textarea>
        </div>
        <div className="mb-3">
          <button type="submit" className="btn btn-outline-success">
            AddTodo
          </button>
        </div>
      </form>
    </>
  );
}
