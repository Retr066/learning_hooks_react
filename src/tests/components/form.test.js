import { shallow } from "enzyme";
import Form from "../../components/Form";

describe("Pruebas en el componente de Form", () => {
  const dispatch = jest.fn();
  let wrapper = shallow(<Form dispatch={dispatch} />);
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<Form dispatch={dispatch} />);
  });

  test("debe renderizar correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("no debe llamar a la function handleFormSubmit", () => {
    const handleFormSubmit = wrapper.find("form").prop("onSubmit");
    handleFormSubmit({ preventDefault() {} });
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(dispatch).not.toHaveBeenCalled();
  });
  test("debe llamar ala function dispatch", () => {
    const title = "Jose";
    const desc = "hacer algo";

    wrapper.find("input").simulate("change", {
      target: {
        name: "title",
        value: title,
      },
    });
    wrapper
      .find("textarea")
      .simulate("change", { target: { name: "desc", value: desc } });
    const handleFormSubmit = wrapper.find("form").prop("onSubmit");
    handleFormSubmit({ preventDefault() {} });
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(expect.any(Object));
    expect(wrapper.find("input").prop("value")).toBe("");
    expect(wrapper.find("textarea").prop("value")).toBe("");
  });
});
