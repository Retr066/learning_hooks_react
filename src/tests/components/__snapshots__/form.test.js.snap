// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente de Form debe renderizar correctamente 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="exampleFormControlInput1"
      >
        Title
      </label>
      <input
        autoComplete="off"
        className="form-control"
        name="title"
        onChange={[Function]}
        placeholder="Ingrese su nombre completo"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="exampleFormControlTextarea1"
      >
        Description
      </label>
      <textarea
        autoComplete="off"
        className="form-control"
        name="desc"
        onChange={[Function]}
        onKeyDown={[Function]}
        rows="3"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <button
        className="btn btn-outline-success"
        type="submit"
      >
        AddTodo
      </button>
    </div>
  </form>
</Fragment>
`;
